tests/Middleware/AuthTokenMiddlewareTest.php

```diff
@@ -306,6 +285,63 @@ public function testShouldNotifyTokenCallback(callable $tokenCallback)
         $this->assertTrue(MiddlewareCallback::$called);
     }
 
+    public function testAddAuthHeadersFromUpdateMetadata()
+    {
+        $authResult = [
+            'authorization' => 'Bearer 1/abcdef1234567890',
+        ];
+
+        $this->mockFetcher->willImplement(UpdateMetadataInterface::class);
+        $this->mockFetcher->updateMetadata(Argument::cetera())
+            ->shouldBeCalledTimes(1)
+            ->willReturn($authResult);
+        $this->mockFetcher->getLastReceivedToken()
+            ->willReturn(['access_token' => '1/abcdef1234567890']);
+
+        $request = new Request('GET', 'http://foo.com');
+
+        $middleware = new AuthTokenMiddleware($this->mockFetcher->reveal());
+        $mock = new MockHandler([new Response(200)]);
+        $callable = $middleware($mock);
+        $callable($request, ['auth' => 'google_auth']);
+    }
+
+    public function testOverlappingAddAuthHeadersFromUpdateMetadata()
+    {
+        $authHeaders = [
+            'authorization' => 'Bearer 1/abcdef1234567890',
+            'x-goog-api-client' => 'extra-value'
+        ];
+
+        $request = new Request('GET', 'http://foo.com');
+        $request->withHeader('x-goog-api-client', 'default-value');
+
+        $this->mockFetcher->willImplement(UpdateMetadataInterface::class);
+        $this->mockFetcher->updateMetadata(Argument::cetera())
+            ->shouldBeCalledTimes(1)
+            ->willReturn($authHeaders);
+        $this->mockFetcher->getLastReceivedToken()
+            ->willReturn(['access_token' => '1/abcdef1234567890']);
+
+        $middleware = new AuthTokenMiddleware($this->mockFetcher->reveal());
+        $mock = new MockHandler([function ($request, $options) use ($authHeaders) {
+            $this->assertEquals($authHeaders['authorization'], $request->getHeaderLine('authorization'));
+            $this->assertArrayHasKey('x-goog-api-client', $request->getHeaders());
+            return new Response(200);
+        }]);
+        $callable = $middleware($mock);
+        $callable($request, ['auth' => 'google_auth']);
```

Comment:
This one is better, since we have assertions in the mock handler, but we have no guarantee that the function in `MockHandler` is called. Something like this could work for this:

```php
$called = false;
$mock = new MockHandler([function ($request, $options) use ($authHeaders, &$called) {
    $this->assertEquals($authHeaders['authorization'], $request->getHeaderLine('authorization'));
    $this->assertArrayHasKey('x-goog-api-client', $request->getHeaders());
    $called = true;
    return new Response(200);
}]);
// and then at the end of the test...
$this->assertTrue($called);
```

Reply:
Thanks for the pointer. Implemented.